### PR TITLE
Implemented Water Supply DTO and Tests

### PR DESCRIPTION
**Pull request by zack-rma**
Implements water supply DTO and appropriate serialization and validation tests.

### CODE REVIEW

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/LookupType.java
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/LookupType.java]: There is already a LookupType class

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/LocationRefType.java
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/LocationRefType.java]: CwmsId should be used instead of this class

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserType.java
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserType.java]: Don't need the `Type` suffix. This is the WaterUser

Comment (adamkorynta) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserType.java]: same with the other classes with `Type` suffix.

Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserType.java]: Addressed in latest commit

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/LocationRefType.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/LocationRefType.java]: Addressed in latest commit

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/LookupType.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/LookupType.java]: Addressed in latest commit

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/watersupply/waterusercontract.json
Comment (MikeNeilson) [inline cwms-data-api/src/test/resources/cwms/cda/data/dto/watersupply/waterusercontract.json]: Just thinking semantically here, the "reference" has more data than the "contract"
I suspect the content is correct but the naming itself is off slightly.

Also is this contract from the contract perspective or the water users' perspective?
Naming wise looks like user perspective.

I'm not vested in either way, but it's not super clear on initial read.

**Review by MikeNeilson — COMMENTED** [review]
I have concerns, but I'm not familiar enough with how this data is used to set it to "request changes."

**Review by MikeNeilson — DISMISSED** [review]

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/watersupply/waterusercontract.json
Comment (zack-rma) [reply cwms-data-api/src/test/resources/cwms/cda/data/dto/watersupply/waterusercontract.json]: Changed naming scheme in recent commit

**Comment by zack-rma** [comment]
Sorry about the rebase, Mike.

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/watersupply/waterusercontract.json
Comment (zack-rma) [reply cwms-data-api/src/test/resources/cwms/cda/data/dto/watersupply/waterusercontract.json]: Renamed to water-user. This contract should be from the contract perspective

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java]: use @JsonProperty(required=true) on the class variable instead of implementing validateInternal

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUser.java
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUser.java]: use @JsonProperty(required=true) on the class variable instead of implementing validateInternal

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java]: use @JsonProperty(required=true) on the class variable instead of implementing validateInternal

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContractRef.java
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContractRef.java]: use @JsonProperty(required=true) on the class variable instead of implementing validateInternal

**Review by adamkorynta — CHANGES_REQUESTED** [review]

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/PumpType.java
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/PumpType.java]: Why vary the enum name from the string name? PUMP IN vs IN

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java]: This pump location is also a stream location, not sure if you would change anything here but perhaps discuss with Bryson.

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java]: Are both of these constructors needed and both are being used?

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUser.java
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUser.java]: Is this constrained to be a project? If so rename the variable to indicate project

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContractRef.java
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContractRef.java]: Is there a reason why you didnt compose this into WaterUserContract?

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java]: this variable name is ambiguous and results in confusing json.
  "contract-id": {
    "office-id": "MVR",
    "name": "TEST_CONTRACT"
  },
  "water-contract": {
    "office-id": "MVR",
    "display-value": "Test Display Value",
    "tooltip": "Test Tooltip",
    "active": true
  },

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java:65]: The generated JSON for these has redundancies. I'm not sure if that is a big deal though.
  "pump-out-location": {
    "pump-location": {
      "office-id": "MVR",
      "name": "PUMP1",
      "latitude": 0,
      "longitude": 0,
      "active": true,
      "public-name": "Test Public Pump Name",
      "long-name": "Test Long Name",
      "description": "Test Description",
      "timezone-name": "UTC",
      "location-type": "Test Location Type",
      "location-kind": "PUMP",
      "nation": "US",
      "state-initial": "NV",
      "county-name": "Clark",
      "nearest-city": "Sparks",
      "horizontal-datum": "WGS84",
      "published-longitude": 0,
      "published-latitude": 0,
      "vertical-datum": "WGS84",
      "elevation": 150,
      "map-label": "Test Map Label",
      "bounding-office-id": "MVR",
      "elevation-units": "m"
    },
    "pump-type": "PUMP_OUT"
  },

**Review by rma-psmorris — CHANGES_REQUESTED** [review]
Questions, not sure if changes are needed.

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java]: Error is thrown during serialization round trip tests when empty constructor is not present

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java]: Will chat with Bryson about that

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/PumpType.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/PumpType.java]: No particular reasoning, can be shortened to just IN, OUT, BELOW

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUser.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUser.java]: Will rename variable

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java]: Will rename waterContract to contractType

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java:65]: I tried to reduce redundancies, but there are still duplicate office IDs. Not sure if it makes a big difference

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContractRef.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContractRef.java]: I believe it was meant to be removed, will refactor to avoid usage

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java]: As long as the pump location is unique to the pump, all should be fine. It is currently designed to store the pump location when the contract is saved.

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java
Comment (rma-rripken) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java]: You can annotate the constructor to let Jackson know which properties go to which ctor args.  We don't do it often but I can show you how

Comment (rma-rripken) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java]: Look at LocationCategory for an example

Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterSupplyPump.java]: Removed empty constructors and tagged input parameters with JSON tags

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/watersupply/waterusercontract.json
Comment (adamkorynta) [inline cwms-data-api/src/test/resources/cwms/cda/data/dto/watersupply/waterusercontract.json]: should be `project-id` for consistency

Comment (zack-rma) [reply cwms-data-api/src/test/resources/cwms/cda/data/dto/watersupply/waterusercontract.json]: Renamed to `project-id`

**Review by adamkorynta — COMMENTED** [review]
Looks good, will wait for other conversations to resolve before adding approval

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java
Comment (rma-psmorris) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java:65]: duplicate offices are expected

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java:50]: Why is this a Date instead of being an Instant?

**Review by adamkorynta — DISMISSED** [review]

**Review by adamkorynta — APPROVED** [review]

**Review by rma-psmorris — COMMENTED** [review]
One question

File: cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/watersupply/WaterUserContract.java:50]: A leftover from before I started transitioning all dates to instants, I must have missed this one. Will update in the controller PR

**Pull request merged**
